src/html5/parser.rs

```diff
@@ -3721,18 +3812,39 @@ impl<'chars> Html5Parser<'chars> {
         }
     }
 
+    fn parser_data(&self) -> ParserData {
+        if self.open_elements.is_empty() {
+            return ParserData {
+                adjusted_node_namespace: HTML_NAMESPACE.to_string(),
+            };
+        }
+
+        let namespace = self
+            .get_adjusted_current_node()
+            .namespace
+            .expect("namespace not found");
+
+        ParserData {
+            adjusted_node_namespace: namespace,
+        }
+    }
+
     /// Fetches the next token from the tokenizer. However, if the token is a text token AND
     /// it starts with one or more whitespaces, the token is split into 2 tokens: the whitespace part
     /// and the remainder.
     fn fetch_next_token(&mut self) -> Token {
         // If there are no tokens to fetch, fetch the next token from the tokenizer
         if self.token_queue.is_empty() {
-            let token = self.tokenizer.next_token().expect("tokenizer error");
+            let token = self
+                .tokenizer
+                .next_token(self.parser_data())
+                .expect("tokenizer error");
 
             if let Token::Text(value) = token {
-                for c in value.chars() {
-                    self.token_queue.push(Token::Text(c.to_string()));
-                }
+                self.token_queue.push(Token::Text(value));
```

Comment:
We can safely push the whole value directly as a single token instead of splitting each character as a separate token. This saves us A LOT of time parsing